### DISCUSSION TITLE
How to display a (settings) dialog before MainWindow (on first run)

### DISCUSSION THREAD

**Discussion by lumip** [AvaloniaUI/Avalonia#7643] [Q&A] [answered]
I am writing an application that requires some configuration on/before the first run. I would therefore like to open the settings dialog automatically when no stored configuration is detected *before showing the main window*. (Since the contents of the main window depend on these settings, there's nothing meaningful to display - so I would rather not display the main window at all than having a bunch of case distinction for settings/no settings in the associated view model).
What would be the best way to achieve this in Avalonia?
So far I have tried:

- 1. Simply creating the settings dialog from the `App.OnFrameworkInitializationCompleted` method and displaying it with `settingsView.ShowDialog`. This fails because `ShowDialog` requires a parent window.
- 2. Assigning my SettingsView instance to `desktop.MainWindow` and subscribing to the `SettingsViewModel.ApplyCommand` which closes the settings dialog and returns the settings - the subscribing action then assigns a properly initialised main window view to `desktop.MainWindow`. This fails because closing the dialog (now being the main window) will exit the application before the proper main window gets set up.
- 3. Same as 2. but preventing the dialog from being closed (using the `Closing` event with an event handler that simply cancels the request) until the proper main window is assigned. This fails: the application still exists after the "Apply" button in my dialog is hit. (The MainWindowViewModel runs through its completey constructor and initialises correctly before then being killed immediately...)

What am I missing to let Avalonia know to "attach" the overall lifetime to the new main window?

Edit: I've put a simplified example that demonstrate the problem in this repo: https://github.com/lumip/AvaloniaDialogFirstMWP

**Comment by ahopper** [comment, marked as answer]
One approach would be to simply display the settings dialog view in the main window i.e it is the main window initially.

**Reply by timunie** [reply]
Yes, or open the SettingsWindow first and on closing start the MainWindow. For sure you will need to perform an additional check on startup.

**Reply by lumip** [reply]
That was my initial thought as well, but I must be doing something wrong because I'm running into the problem of the application terminating after the settings dialog is closed in that case, so how would I do this properly?

**Reply by workgroupengineering** [reply]
Have you tried to replace the main windows before closing the setting dialog?

**Reply by lumip** [reply]
I have. The problem is that I have to do this from a subscription to the dialog's `ApplyCommand` and it seems to be not possible to re-assign the MainWindow from inside there (simply nothing happens). (I have confirmed that the subscribing action/lambda completes and definitely does the assignment of my main window instance to `desktop.MainWindow`, so it's not a problem of that code not getting executed..)

**Reply by timunie** [reply]
Can you provide a sample via GitHub? I think this could help me try to get this done. 

Ideas I have
- change the way the App shuts down
- cancel shutdown
- maybe change some order of execution
- use Dispatcher
- use ContentDialog instead
- ...

But to see which one helps in your case I think a sample would be helpful.

**Reply by lumip** [reply]
I've put a simplified example that demonstrates the problem in this repo: https://github.com/lumip/AvaloniaDialogFirstMWP
Thanks for having a look!

**Reply by workgroupengineering** [reply]
```csharp
                dialog.ViewModel!.ApplyCommand
                    /*.ObserveOn(RxApp.MainThreadScheduler).SubscribeOn(RxApp.MainThreadScheduler)*/
                    .Subscribe(result =>
                {
                    var mw = new MainWindow
                    {
                        DataContext = new MainWindowViewModel(result),
                    };
                    desktop.MainWindow = mw;
                    mw.Show();
                    dialog.Close();
                });
```

**Reply by lumip** [reply]
Alright, that works. What a stupid little thing to miss. Thanks a lot everyone!
